### PR TITLE
Add goleak checking to tests

### PR DESCRIPTION
**Pull request by brawndou**
This adds goleak testing across all tests in package tally and fixes goroutine leaks in scope related tests.

### CODE REVIEW

File: scope_registry_test.go [hunk 1]
```diff
@@ -143,10 +143,16 @@ func TestNewTestStatsReporterManyScopes(t *testing.T) {
...
-			hello.Inc(1)
+			select {
+			case <-quit:
+				return
+			default:
+				hello := root.Tagged(map[string]string{"a": "b"}).Counter("hello")
+				hello.Inc(1)
+			}
 		}
 	}()
```
Comment (mway) [inline scope_registry_test.go:157]: Even though we have partial synchronization here (send/recv across an unbuffered channel), it only covers the recv itself and the function could be preempted before exiting (particularly if more logic is added in the future), so we should still ensure that it exits before the test completes. Even just something like:

```go
var (
  root = newRootScope(/* ... */)
  quit = make(chan bool)
  done = make(chan struct{})
)

go func() {
  defer close(done)

  for {
    select {
    case <-quit:
      return
    default:
      // ...
    }
  }
}()

// ...

var c Counter
for {
  // ...
}

<-done
```

(Applies both here and the snapshot case below.)

File: scope_registry_test.go [hunk 2]
```diff
@@ -143,10 +143,16 @@ func TestNewTestStatsReporterManyScopes(t *testing.T) {
 
 func TestForEachScopeConcurrent(t *testing.T) {
 	root := newRootScope(ScopeOptions{Prefix: "", Tags: nil}, 0)
+	quit := make(chan bool)
```
Comment (mway) [inline scope_registry_test.go:146]: nit: style

```suggestion
	var (
		root = newRootScope(ScopeOptions{Prefix: "", Tags: nil}, 0)
		quit = make(chan bool)
	)
```

File: scope_test.go [hunk 3]
```diff
@@ -1167,16 +1172,23 @@ func TestSnapshot(t *testing.T) {
 
 func TestSnapshotConcurrent(t *testing.T) {
 	scope := NewTestScope("", nil)
+	quit := make(chan bool)
```
Comment (mway) [inline scope_test.go:1175]: nit: style

```suggestion
	var (
		scope = NewTestScope("", nil)
		quit  = make(chan bool)
	)
```

File: scope_registry_test.go (hunk 1 above)
Comment (brawndou) [reply scope_registry_test.go]: done

File: scope_registry_test.go (hunk 2 above)
Comment (brawndou) [reply scope_registry_test.go]: done

File: scope_test.go [hunk 4]
```diff
@@ -353,6 +354,10 @@ func (r *testStatsReporter) Flush() {
 	atomic.AddInt32(&r.flushes, 1)
 }
 
+func TestMain(m *testing.M) {
+	goleak.VerifyTestMain(m)
+}
```
Comment (mway) [inline scope_test.go:359]: let's put this at the top of the file as the first test since it's `TestMain`

**Review by mway — APPROVED** [review]
lgtm modulo moving `TestMain` up top, thanks!

**Pull request merged**